test: Replace MetalLB with dummyLB and re-enable LoadBalancer test

brb opened the pull request:
This PR re-enables LoadBalancer service tests, and switches from MetalLB to dummylb \[1\]. The main reason for switching is that the former was operating in L2 mode which interfered with the vbox VM bridge resulting in the test flakes.

The new lb only assigns LoadBalancerIP addr to a service (specified by a user in `.spec.loadBalancerIP`). Therefore, making the LB IP addr reachable from other nodes is up to a user. This can be achieved by installing a route on a client node (as we do in the test).

Also, refactor `IP{Add,Del}Route` and `ExecIn{FirstPod,HostNetNS}` helpers (see commit messages).

\[1\]: https://github.com/cilium/dummylb.

Fix #10763

brb committed:
test: Re-enable LoadBalancer tests

This commit re-enables LoadBalancer service tests, and switches from
MetalLB to dummylb [1]. The main reason for switching is that the former
was operating in L2 mode which interfered with the vbox VM bridge
resulting in the test flakes.

The new lb only assigns LoadBalancerIP addr to a service (specified by a
user in .spec.loadBalancerIP). Therefore, making the LB IP addr reachable
from other nodes is up to a user. This can be achieved by installing a
route on a client node (as we do in the test).

[1]: https://github.com/cilium/dummylb.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Refactor IP{Add,Del}Route helpers

- Invoke cmds for a given node instead of returning cmd itself - reduces
  boilerplate by quite a bit.
- Changes method names form to verb + noun.

Suggested-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

christarazi commented on test/helpers/kubectl.go:
```diff
@@ -4102,3 +4102,30 @@ func (kub *Kubectl) GetDNSProxyPort(ciliumPod string) int {
+
+	res, err := kub.ExecInHostNetNS(context.TODO(), nodeName, cmd)
+	gomega.Expect(res).To(CMDSuccess())
+	gomega.Expect(err).ToNot(gomega.HaveOccurred(), "Cannot exec %s in node %s", cmd, nodeName)
```
I think this refactor is a good one overall. 

The issue I have with this is what I've highlighted. In general, we should avoid doing expectations inside helper functions. The rationale is that it screws up the stacktrace when the test fails. As an example here, if `AddIPRoute` fails, the test output will point at this funciton, but not *where* `AddIPRoute` was called. To workaround that, we can use `ExpectWithOffset`. However, I think we should avoid using the offset variation when possible, to simplify thinking about the case when there's potentially nested calls. (I've also seen variations of our helpers accept an `offset` to pass into `ExpectWithOffset`, but I feel this is a bit too much. We should strive for simplicity instead.)

I would propose to return the `res` and `err` instead, and let the caller of these helpers perform the expectations. I understand trying to reduce the boilerplate of doing expectations on helper functions, but I think having consistent and readable stacktraces when a test fails is more valuable than "elegant" test code, if you will :).

brb replied:
Yeah, good points re the offset. Anyway, please take a look at another approach (two last commits).

christarazi commented on test/k8sT/Services.go:
```diff
@@ -1485,28 +1485,38 @@ var _ = Describe("K8sServicesTest", func() {
+							cmd := helpers.IPAddRoute(net.ParseIP(lbIP), net.ParseIP(k8s1IP), false)
+							res, err := kubectl.ExecInHostNetNS(context.TODO(), outsideNodeName, cmd)
+							ExpectWithOffset(1, err).Should(BeNil(), "Cannot exec in outside node %s", outsideNodeName)
+							ExpectWithOffset(1, res).Should(helpers.CMDSuccess(),
```
I think these should be `Expect` instead of `ExpectWithOffset`. Generally, you would use `ExpectWithOffset` inside a helper function, but inside an `It()` you want to use `Expect`, so that the stacktraces point to exactly where in the code the test failed.

christarazi reviewed (changes requested):
A few areas to look into wrt to CI semantics

brb committed:
test: Make ExecInFirstPod() to return a single value

Instead of returning (*CmdRes, error) make the method to return *CmdRes
by setting error in the CmdRes.err field.

This makes assertions of functions which depend on the method simpler.
A notable example is ExecInHostNetNS() which previously required to
assertions.

Finally, extend the *CMDRes success matcher by adding a check whether
CmdRes.err is not nil. This will help us to catch bugs when CmdRes was
considered as successful, although an error was set.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on test/k8sT/manifests/demo_ds.yaml:
```diff
@@ -209,6 +209,7 @@ metadata:
   name: test-lb
 spec:
   type: LoadBalancer
+  loadBalancerIP: 192.168.36.240
```
this field should be populated by the LB. Also, different infrastructures will have different IPs.

brb replied:
dummyLB will set that IP in the status. The IP is set from the same range as we used before with MetalLB.

brb replied:
https://github.com/cilium/dummylb/blob/master/main.go#L39

aanm reviewed (changes requested):
I didn't pay too much attention to the CI code.

christarazi commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -756,9 +756,11 @@ func testPodHTTPToOutside(kubectl *helpers.Kubectl, outsideURL string, expectNod
 			// Make pods reachable from the host which doesn't run Cilium
-			kubectl.AddIPRoute(helpers.GetNodeWithoutCilium(), podIP, hostIP, false)
+			kubectl.AddIPRoute(helpers.GetNodeWithoutCilium(), podIP, hostIP, false).
+				ExpectSuccess("Failed to add ip route")
```
I'm not sure what kind of stacktrace we'll get in this case if it fails because `ExpectSuccess` does an expect with offset of 1 and we're already inside a function that's called within an `It()`.

Will we see:

```
testPodHTTPToOutside:760
It(...)
Context(...)
...
```

Or

```
It(...)
Context(...)
```

It's worth making this fail on purpose, and see what kind of stacktrace you get.

This scenario here  is precisely why I would like to avoid all uses of `ExpectWithOffset` and helper functions that perform expectations. Rather just return their error (or CmdRes), and let the caller, who has the context of what the stacktrace should look like, decide if there's a need for an offset or not.

joestringer replied:
I agree on checking what kind of stacktraces we get.

I'm guessing that @christarazi your take on reducing/removing `ExpectWithOffset()` and helper expectations is that these functions should instead return an error in those conditions then allow the caller to assert properly. That way, the specific error can be identified, but we also have a meaningful point in the actual test code to highlight which part failed?

joestringer replied:
The other aspect to this is that I think this is a known shortfall of the current way we have been using Ginkgo, and it'd be nice to figure out a reasonable line for @brb here such that we're able to unblock the users of the changes in this PR without forcing this PR to solve error handling generically in CI.

christarazi replied:
Yes @joestringer that is what I am trying to get at :). In summary, it is better to return the errors, rather than make expectations inside the helper.

I should've clarified that, yes. We shouldn't feel the need to refactor the entire codebase in one go right now. Rather, fix up the usages that were already modified in this PR. Specifically, after understanding what the stacktrace looks, we either keep how it is now with a call to the helper with an`ExpectSuccess`, or we avoid using `ExpectSuccess` and just perform the expectation explicitly (with the correct offset) on the return value of these helpers. I hope that clarifies things for @brb.

brb replied:
> It's worth making this fail on purpose, and see what kind of stacktrace you get.

For this case I get:
```
K8sDatapathConfig
/home/brb/sandbox/gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:478
  Check BPF masquerading with ip-masq-agent
  /home/brb/sandbox/gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:478
    DirectRouting [It]
    /home/brb/sandbox/gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:514

    Failed to add ip route
    Expected command: kubectl exec -n kube-system log-gatherer-drzvk -- ip route add asdf via 192.168.36.12
    To succeed, but it failed:
    Exitcode: 1
    Err: Process exited with status 1
    Stdout:

    Stderr:
         Error: any valid prefix is expected rather than "asdf".
         command terminated with exit code 1



    /home/brb/sandbox/gopath/src/github.com/cilium/cilium/test/k8sT/DatapathConfiguration.go:760
```

ACK on making helpers to return an error and asserting in `Context/It` bodies instead. However, the failed expectations, which my PR touches, are clear enough to see what exactly failed (we don't call thousands of helpers from inside `Context/It`). Therefore, I suggest to refactor helpers in a separate PR (I can do in a follow-up).

christarazi commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -756,9 +756,11 @@ func testPodHTTPToOutside(kubectl *helpers.Kubectl, outsideURL string, expectNod
+				ExpectSuccess("Failed to add ip route")
 			defer func() {
-				kubectl.DelIPRoute(helpers.GetNodeWithoutCilium(), podIP, hostIP)
+				kubectl.DelIPRoute(helpers.GetNodeWithoutCilium(), podIP, hostIP).
+					ExpectSuccess("Failed to del ip route")
```
I believe because this is inside a function, it'll need an extra offset. Worth making it fail to see.

christarazi reviewed (commented):
I like the improvement of consolidating the returns and making it concise. A few points below. I also assume you'll squash the related refactors into one commit?

joestringer reviewed (changes requested):
GKE seems to have a few failures particularly in tests modified by this PR. How do we want to handle this?
https://jenkins.cilium.io/job/Cilium-PR-K8s-GKE/2220/#showFailuresLink

brb committed:
test: Do not provision LB SVC with IP addr on GKE

Otherwise, GKE LB might fail to assign the IP addr to other LB svcs
which results in the following failure:

  <*errors.errorString | 0xc0002a6410>: {
              s: "could not get service LoadBalancer IP addr: 30s
              timeout expired",
  }

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Add K8sUpdate on GKE to the quarantine

Currently, it's broken, and it prevents other tests from being invoked.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
